tests/python/pants_test/engine/test_fs.py

```diff
@@ -708,6 +715,95 @@ def test_nonexistent_filename_globs(self) -> None:
         subset_digest = self.request_single_product(Digest, subset_input)
         assert subset_snapshot.directory_digest == subset_digest
 
+    def test_file_content_invalidated(self) -> None:
+        """Test that we can update files and have the native engine invalidate previous operations
+        on those files."""
+
+        with self.mk_project_tree() as project_tree:
+            scheduler = self.mk_scheduler(rules=create_fs_rules(), project_tree=project_tree)
+            fname = "4.txt"
+            new_data = "rouf"
+            # read the original file so we have a cached value.
+            self.read_file_content(scheduler, [fname])
+            path_to_fname = os.path.join(project_tree.build_root, fname)
+            with open(path_to_fname, "w") as f:
+                f.write(new_data)
+            for i in range(4):
```

Comment:
Since this pattern is used three times here it might be good to extract it... maybe:
```
def try_with_backoff(thing_to_try: Callable[[], bool]) -> bool:
  for i in range(4):
    time.sleep(0.1 * i)
    if thing_to_try:
      return true
  return false

def test_*():
  ..
  def my_thing():
    ..
    return true_if_it_worked
  assert try_with_backoff(my_thing)
```